### DISCUSSION TITLE
Separate authentication from user profile retrieval

### DISCUSSION THREAD

**Discussion by damien-git** [vufind-org/vufind#2585] [Q&A]
I am trying to resolve a particular issue with authentication: we need to give users a choice of 2 authentications: FOLIO or SAML. I started implementing a custom driver for SAML using simplesamlphp, but for the purpose of this issue, we can just assume it is Shibboleth. When users authenticate with FOLIO, they provide a username and password. When they authenticate with SAML, we are only getting the username back from the IdP and want to get all the other user profile information from FOLIO. When a user is authenticated, no matter the method, we want to use a special FOLIO user with permissions to retrieve user-related information like firstname or checked out items.

I have tried using ChoiceAuth, and set Folio.ini with okapi_login = true. This works for users authenticating with FOLIO, but users authenticating with SAML are being prompted for their catalog username and password. How could I make this work in Vufind ? There is only one Folio.ini, so I can't use a different value of okapi_login for each method, and also it looks like authentication, user profile retrieval and user catalog information (like checked out items and fines) are all tied...

**Comment by demiankatz** [comment]
@damien-git, the intended design here is that all authentication methods need to correctly populate the cat_username and cat_password fields in the user table of the database. These are the credentials used to obtain data from the ILS. As long as your authentication populates these fields with values that will work with the ILS driver, everything should work seamlessly regardless of authentication method chosen. Does that help, or is there a reason that you cannot populate the fields consistently in your different authentication scenarios?

Based on your description, it sounds to me like you may want to set okapi_login to false in Folio.ini to make it easier to populate appropriate attributes. See also current work in progress on #2570 for a new feature that will make things more secure if you wish to use weaker credentials in the cat_username / cat_password fields to simplify the integration.

**Reply by damien-git** [reply]
@demiankatz Thanks for your reply. Our IdP will not return anything for cat_username and cat_password. The username is actually the same with SAML and FOLIO for people authenticated with SAML, but the password is an issue. We don't want to store any password in the database. There is a single username/password used to get information from FOLIO, which we want to use to get profile information in all cases.
Also, if we have okapi_login=false, how could users still authenticate with FOLIO (the users who don't use SAML) ? I think okapi_login needs to be true for them.

**Reply by demiankatz** [reply]
Yes, good points. If you don't want to store passwords in the database, I would suggest this approach:

1.) In FOLIO.ini, set okapi_login = false, and set username_field and password_field appropriately. Until #2570 is done, I'd suggest choosing two distinct user attributes that you can pull from authentication details to prevent the possibility of user-spoofing. Once #2570 is done, you should be able to safely match on username alone.

2.) Instead of using the standard ILS authentication, build a local custom Auth module for FOLIO authentication. This local auth module will use Okapi authentication to validate the user, but it will act independently of the configured ILS driver. I think you could build this by starting with the VuFind\Auth\ILS class and ripping out everything except the authenticate, handleLogin, processILSUser and getUsernameField methods. Large parts of some of these methods could be removed/simplified (e.g. removing support for email-based authentication, which is not relevant to your case). Instead of using $this->getCatalog() to access the VuFind\ILS\Connection object to proxy driver methods, you could locally construct an instance of the FOLIO driver, used only by the authentication module, and call its methods directly. To configure this driver instance, you could load the FOLIO.ini file from disk, change a few settings in the resulting array (mainly okapi_login and anything related to that), and then setConfig the results. This should allow you to validate users using Okapi just in the context of the authentication module's handleLogin method, and then save necessary details in the user table through processILSUser so that the rest of the code can use the globally-configured driver to access ILS functions without using the Okapi login.

I realize that building custom local code is not necessarily the most desirable solution to a problem, but I think this proposal achieves your goals while mostly leveraging existing code -- just in a slightly non-standard configuration. Does that make any sense? If I'm not expressing myself clearly, or if you try it and run into trouble, I'm happy to provide more details, example code, etc.

**Reply by damien-git** [reply]
Thanks. I think 2) should work, I will try it. I was hoping there would be an easier solution I missed, but I guess not. Regarding 1), I am not sure how user spoofing would be possible in our case. There are FOLIO users matching all users available through SSO, so a username conflict between the 2 authentication methods is not possible.

**Reply by demiankatz** [reply]
@damien-git, under normal circumstances, username spoofing should not be possible, but under certain conditions (e.g. a temporary ILS outage causing ILS authentication to fail), VuFind could potentially display the ILS login form and provide an opportunity for users to enter inappropriate credentials. #2570 adds a setting which will completely put an end to that possibility (and the changes are pretty simple to backport if necessary), but without those changes in place, it's something to be cautious about.

**Reply by damien-git** [reply]
@demiankatz I implemented your solution, but it did not work: users authenticating with SAML are still getting a catalog login form. Looking at AbstractBase#catalogLogin, I don't see how this could work...
Edit: I guess catalogLogin could call storedCatalogLogin() which would call patronLogin(), which would just fetch the user because okapi_login is false. I'll have to look more into it. I have avoided using "$driverManager->get('Folio')" to create the Folio driver in the custom auth driver so that the ILS authenticator is different (which was not simple), so that should not be an issue...

**Reply by demiankatz** [reply]
@damien-git, is the SAML authentication correctly populating the cat_username and cat_password fields in the user table with values that can be used to look up the user? The catalog login form should only appear when those values are missing or incorrect.

**Reply by damien-git** [reply]
I got it to work eventually, I had some bugs. I still wonder if I can safely use "$driverManager->get('Folio')" to create the Folio driver in the custom Folio auth constructor. Would there be a risk that the custom config overrides the general one ? My initialization looks like this:
```
        $this->driver = $driverManager->get('Folio');
        $config = $configReader->get('Folio');
        $driverConfig = is_object($config) ? $config->toArray() : [];
        $driverConfig['User']['okapi_login'] = true;
        $this->driver->setConfig($driverConfig);
        $this->driver->init();
```

**Reply by demiankatz** [reply]
@damien-git, it would probably be a good idea to clone the driver after fetching it from the driverManager, to ensure that you have a distinct instance from the one that will be used in other contexts.

**Reply by damien-git** [reply]
I was trying to call the constructor directly, but that's made more complicated by the various traits that need to be initialized. How do you clone this driver object ?

**Reply by demiankatz** [reply]
@damien-git, you can just use the PHP clone operator: `$this->driver = clone $driverManager->get('Folio');`

**Reply by damien-git** [reply]
Thanks, I'm going to use that.
